File: api/src/main/java/com/gkh/syntheticmonitor/model/ReportDetail.java
package com.gkh.syntheticmonitor.model;


import javax.persistence.*;

import com.fasterxml.jackson.annotation.JsonGetter;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Builder;
import lombok.Data;


@Entity
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Access(AccessType.FIELD)
public class ReportDetail {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;
	private String name;
	private String details;
	private String type;
	private String status;

	private long responseTime;
	private long maximumResponseThreshold;

	private transient boolean responseTimeUnderMax;
	private transient boolean statusSuccess;

	private String expectedStatus;

	private transient String content;

	@ManyToOne
	@JsonIgnore
	private Report parent;

	@Column
	@Access(AccessType.PROPERTY)
	public boolean isStatusSuccess() {
		return this.isStatusCodeMatching() & this.isResponseTimeUnderMax();
	}


	@Column
	@Access(AccessType.PROPERTY)
	public boolean isResponseTimeUnderMax() {
		return this.responseTime <= this.maximumResponseThreshold;
	}

	@JsonGetter
	public boolean isStatusCodeMatching() {
		return this.status.equals(this.expectedStatus);
	}

}

File: api/src/main/kotlin/com/gkh/syntheticmonitor/model/ReportDetail.kt
package com.gkh.syntheticmonitor.model

import com.fasterxml.jackson.annotation.JsonGetter
import com.fasterxml.jackson.annotation.JsonIgnore
import lombok.AllArgsConstructor
import lombok.Builder
import lombok.Data
import lombok.NoArgsConstructor
import javax.persistence.*

@Entity
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Access(AccessType.FIELD)
class ReportDetail(@Id @GeneratedValue(strategy = GenerationType.AUTO)  val id: Long? = null,
                   var name: String? = null,
                   var details: String? = null,
                   var type: String? = null,
                   var status: String? = null,
                   var responseTime: Long = 0,
                   var maximumResponseThreshold: Long = 0,
                   var expectedStatus: String? = null,
                   @Transient val content: String? = null,
                   @ManyToOne @JsonIgnore val parent: Report? = null) {

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var isResponseTimeUnderMax = false
        get() = responseTime <= maximumResponseThreshold

    @get:Access(AccessType.PROPERTY) @get:Column @Transient
    var isStatusSuccess = false
        get() = isStatusCodeMatching and this.isResponseTimeUnderMax

    @get:JsonGetter
    val isStatusCodeMatching: Boolean
        get() = status == expectedStatus
}